### ISSUE TITLE
Controller mapping "SCS.3d MIDI 1" error on launch after upgrade to mixxx 2.4

### ISSUE THREAD

**Issue by sysdbugfactory** [mixxxdj/mixxx#13133]
### Bug Description

Been using mixxx for a number of years with this setup and recently upgraded to 2.4

Upon launch an error message is raised stating:

> The mapping for your controller "SCS.3d MIDI 1" is not working properly.
> 
> The script code needs to be fixed.
> You can ignore this error for this session but you may experience erratic behavior.
> Try to recover by resetting your controller.

details are:

> Uncaught exception: <passed code>:1: TypeError: Property 'incomingData' of object function StantonSCS3d() { [native  ode] } is not a function
> Backtrace: @:1
> @:1

The Retry button loops on this error.

The Rgnore button allows mixxx to start and the controller seems to mostly work but is missing some features.

The expected behavior is no error and the controller working as it used to in previous mixxx version.

### Version

2.4.0

### OS

Linux (Manjaro)

**Comment by sysdbugfactory** [comment]
I should that the controller is a [Stanton SCS.3d "DaScratch"](https://github.com/mixxxdj/mixxx/wiki/Stanton%20SCS.3d).

This is a certified mappings, so it is unexpected that it would raise an error.

Hopefully it's not a big issue but I do not grasp enough to be able to fix it myself.

**Comment by Swiftb0y** [comment]
I don't think so. I think the hardware unexpectedly sends a sysex message which is automatically dispatched to `StantonSCS3d.incomingData`. That function however does not exist because the mapping doesn't expect to receive any SysEx data from the hardware. The easy fix would be to add a dummy implementation of `StantonSCS3d.incomingData` so engine doesn't complain. It doesn't explain why the controller sends that in the first place though.

**Comment by git-developer** [comment]
> It doesn't explain why the controller sends that in the first place though.

This is caused by a [sysex command sent from `init()`](https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/res/controllers/Stanton-SCS3d-scripts.js#L229) requesting the firmware version. [statusResponse()](https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/res/controllers/Stanton-SCS3d-scripts.js#L229) is the intended handler, configured [in the XML file](https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/res/controllers/Stanton%20SCS.3d.midi.xml#L19) for status `0xf0`.

Maybe a status of `0xf0` in the XML was enough to route the message to the configured handler in 2.3, and this behavior changed from 2.4?

Workaround is to add 

```js
StantonSCS3d.incomingData = StantonSCS3d.statusResponse;
```

after declaration of `StantonSCS3d.statusResponse`.

**Comment by git-developer** [comment]
Related: #11536, #12824

**Comment by git-developer** [comment]
Problem was introduced with 2.4.1 (does not occur on 2.4.0 and below).

**Comment by Swiftb0y** [comment]
> This is caused by a [sysex command sent from init()](https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/res/controllers/Stanton-SCS3d-scripts.js#L229) requesting the firmware version.

Oh yeah. I missed that. 

> Workaround is to add
>
>> StantonSCS3d.incomingData = StantonSCS3d.statusResponse;
>
> after declaration of StantonSCS3d.statusResponse.

Yup. Do you want to open a PR?

**Comment by git-developer** [comment]
I can do, but I'd like to understand better what's happening before. What I found out so far:

#### 2.3.6
Lookup for an input function by key `[0xF0, 0xFF]` and function call:
https://github.com/mixxxdj/mixxx/blob/691596c177a80d2420a29c2d9273364921d9a21e/src/controllers/midi/midicontroller.cpp#L472
https://github.com/mixxxdj/mixxx/blob/691596c177a80d2420a29c2d9273364921d9a21e/src/controllers/midi/midicontroller.cpp#L490-L493

#### 2.5.0
Declaration of a function `incomingData()` and call of all (really? can't believe that) input functions:
https://github.com/mixxxdj/mixxx/blob/23a9c9a0526c28f53b218b000ee006a49a15116f/src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp#L162-L165

https://github.com/mixxxdj/mixxx/blob/23a9c9a0526c28f53b218b000ee006a49a15116f/src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp#L208-L210

**Comment by git-developer** [comment]
I didn't find out yet where (and why) the difference is made between dispatching of data messages and sysex messages. Nonetheless a PR (draft for now) is open containing a quick fix.

Not sure if more fixes are required. MIDI SysEx has status byte `0xF0`. Some mappings use other system messages, all beginning with `0xF`. Some of them are used in XML mappings:

```xml
$ grep -Hirn 'status>0xf' res/controllers/*.xml
res/controllers/Reloop Beatpad.midi.xml:1987:                <status>0xF0</status>
res/controllers/Roland_DJ-505.midi.xml:3591:                <status>0xFA</status>
res/controllers/Roland_DJ-505.midi.xml:3600:                <status>0xFC</status>
res/controllers/Stanton SCS.1d.midi.xml:18:                    <status>0xf0</status>
res/controllers/Stanton SCS.1d.midi.xml:358:                    <status>0xf9</status>
res/controllers/Stanton SCS.3d.midi.xml:17:                <status>0xf0</status>
```

It might be that these messages are affected, too.

**Comment by git-developer** [comment]
I found the missing puzzle pieces.

[`PortMidiController::poll`](https://github.com/mixxxdj/mixxx/blob/d09d9c55e49e64870584768950075adfb0df1256/src/controllers/midi/portmidicontroller.cpp#L118) checks for SysEx (`0xF0` status byte) and delegates to one of:

https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/src/controllers/midi/midicontroller.cpp#L249-L252

https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/src/controllers/midi/midicontroller.cpp#L572

which in turn delegate to one of

https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/src/controllers/midi/midicontroller.cpp#L298-L302

https://github.com/mixxxdj/mixxx/blob/88aa50e06376e1b2b5a3a3435a3985d771e1c278/src/controllers/midi/midicontroller.cpp#L602-L604

The first `processInputMapping()` (scalar values) dispatches to user-configured bindings.
The second (byte array) does not, but takes JS function `incomingData()` for granted. No default implementation is supplied.

---

My conclusions:
1. The proposed quick fix is OK for this issue.
2. Other system messages than SysEx (e.g. status byte `0xF9`, `0xFA`, `0xFC`) are not related.
3. We should investigate in #11536 how to deal with the problem in general:
    1. Either make `processInputMapping(byte[])` configurable (as in 2.3)
    2. Or make `incomingData()` mandatory, document this, supply an empty default implementation and change the existing mappings by renaming the custom dispatcher function and remove the SysEx binding (`<status>F0</status>`) from XML.
4. #12824 is a duplicate of #11536.

WDYT?

**Comment by Swiftb0y** [comment]
Assuming its reasonably feasible (<100 lines) lets make it configurable again.

**Comment by git-developer** [comment]
I can confirm that this issue does no longer occur in current 2.6 nightly, so it may be closed.

Furthemore I analyzed how custom sysex handlers were dropped in 2.4 and outlined an [idea](https://github.com/mixxxdj/mixxx/issues/11536#issuecomment-2673512381) to bring them back. Implementation should be straightforward but unfortunately exceed my spare C++ skills.

**Comment by daschuer** [comment]
Than let's close this, please reopen if this is still an issue with 2.5.1
